ostd/src/sync/rcu/mod.rs

```diff
@@ -45,101 +46,244 @@ pub use owner_ptr::OwnerPtr;
+        RcuReadGuard {
+            obj_ptr: self.ptr.load(Acquire),
+            rcu: self,
+            _inner_guard: guard,
+        }
+    }
+}
+
+impl<P: OwnerPtr + Send, const MAYBE_UNINIT: bool> Rcu_<P, MAYBE_UNINIT> {
     /// Replaces the current pointer with a new pointer.
-    pub fn replace(&self, new_ptr: P) {
-        let new_ptr = <P as OwnerPtr>::into_raw(new_ptr) as *mut _;
-        let old_ptr = {
-            let old_raw_ptr = self.ptr.swap(new_ptr, AcqRel);
-            // SAFETY: It is valid because it was previously returned by `into_raw`.
-            unsafe { <P as OwnerPtr>::from_raw(old_raw_ptr) }
-        };
-
-        let rcu_monitor = RCU_MONITOR.get().unwrap().lock();
-        rcu_monitor.after_grace_period(move || {
-            drop(old_ptr);
-        });
+    ///
+    /// This function updates the pointer to the new pointer regardless of the
+    /// original pointer. If the original pointer is not NULL, it will be
+    /// dropped after the grace period.
+    pub fn update(&self, new_ptr: P) {
+        let new_ptr = <P as OwnerPtr>::into_raw(new_ptr).cast_mut();
+        let old_raw_ptr = self.ptr.swap(new_ptr, AcqRel);
+
+        if let Some(p) = NonNull::new(old_raw_ptr) {
+            // SAFETY: It was previously returned by `into_raw`.
+            unsafe { delay_drop::<P>(p) };
+        }
     }
 }
 
-/// A guard that allows read-only access to the data protected by the RCU
-/// mechanism.
-///
-/// Note that the data read can be outdated if the data is updated by another
-/// task after acquiring the guard.
-pub struct RcuReadGuard<'a, P: OwnerPtr> {
-    obj: &'a <P as OwnerPtr>::Target,
-    _rcu: &'a Rcu<P>,
+/// A guard that allows read-only access to the initialized data protected
+/// by the RCU mechanism.
+pub struct RcuReadGuard<'a, P: OwnerPtr, const MAYBE_UNINIT: bool> {
+    /// If maybe uninitialized, the pointer can be NULL.
+    obj_ptr: *mut <P as OwnerPtr>::Target,
+    rcu: &'a Rcu_<P, MAYBE_UNINIT>,
     _inner_guard: DisabledPreemptGuard,
 }
 
-impl<P: OwnerPtr> Deref for RcuReadGuard<'_, P> {
+// Initialized RCU guard can be dereferenced.
+impl<P: OwnerPtr> Deref for RcuReadGuard<'_, P, false> {
     type Target = <P as OwnerPtr>::Target;
 
     fn deref(&self) -> &Self::Target {
-        self.obj
+        // SAFETY: Since the preemption is disabled, the pointer is valid
+        // because other writers won't update the pointer until this task
+        // passes the quiescent state.
+        // And this pointer is not NULL.
+        unsafe { &*self.obj_ptr }
     }
 }
 
+// Maybe uninitialized RCU guard can be dereferenced after checking.
+impl<'a, P: OwnerPtr> RcuReadGuard<'a, P, true> {
+    /// Get the initialized read guard.
+    ///
+    /// If the RCU object is not initialized, this function will return
+    /// [`Err`] with itself. Otherwise a dereferenceable read guard will be
+    /// returned.
+    pub fn check(self) -> Result<RcuReadGuard<'a, P, false>, Self> {
```

Comment:
> I suppose that try_get would make a better name

Agreed.

> And the return type should be changed to Option<_>.

No the return type should still be `Result`. If the caller can still preserve the guard on the error path it can initialize it without re-acquiring a read guard.